src/tests/end2end/CopyTests.cpp

```diff
@@ -0,0 +1,502 @@
+
+            nxt::CommandBuffer commands[2];
+
+            nxt::Buffer uploadBuffer = utils::CreateFrozenBufferFromData(device, textureData.data(), static_cast<uint32_t>(sizeof(RGBA8) * textureData.size()), nxt::BufferUsageBit::TransferSrc);
+
+            commands[0] = device.CreateCommandBufferBuilder()
+                .TransitionTextureUsage(texture, nxt::TextureUsageBit::TransferDst)
+                .CopyBufferToTexture(uploadBuffer, 0, rowPitch, texture, 0, 0, 0, width, height, 1, textureSpec.level)
+                .GetResult();
+
+            nxt::Buffer buffer = device.CreateBufferBuilder()
+                .SetSize(bufferSpec.size)
+                .SetAllowedUsage(nxt::BufferUsageBit::TransferSrc | nxt::BufferUsageBit::TransferDst)
+                .SetInitialUsage(nxt::BufferUsageBit::TransferDst)
+                .GetResult();
+            std::vector<RGBA8> emptyData(bufferSpec.size / kBytesPerTexel);
+            buffer.SetSubData(0, static_cast<uint32_t>(emptyData.size()), reinterpret_cast<const uint32_t*>(emptyData.data()));
+
+            commands[1] = device.CreateCommandBufferBuilder()
+                .TransitionTextureUsage(texture, nxt::TextureUsageBit::TransferSrc)
+                .TransitionBufferUsage(buffer, nxt::BufferUsageBit::TransferDst)
+                .CopyTextureToBuffer(texture, textureSpec.x, textureSpec.y, 0, textureSpec.copyWidth, textureSpec.copyHeight, 1, textureSpec.level, buffer, bufferSpec.offset, bufferSpec.rowPitch)
+                .GetResult();
+
+            queue.Submit(2, commands);
+
+            std::vector<RGBA8> expected(bufferSpec.rowPitch / kBytesPerTexel * (textureSpec.copyHeight - 1) + textureSpec.copyWidth);
+            PackTextureData(
+                &textureData[textureSpec.x + textureSpec.y * (rowPitch / kBytesPerTexel)],
+                textureSpec.copyWidth,
+                textureSpec.copyHeight,
+                rowPitch / kBytesPerTexel,
+                expected.data(),
+                bufferSpec.rowPitch / kBytesPerTexel);
+
+            EXPECT_BUFFER_U32_RANGE_EQ(reinterpret_cast<const uint32_t*>(expected.data()), buffer, bufferSpec.offset, static_cast<uint32_t>(expected.size())) <<
+                "Texture to Buffer copy failed copying region [(" << textureSpec.x << ", " << textureSpec.y << "), (" << textureSpec.x + textureSpec.copyWidth << ", " << textureSpec.y + textureSpec.copyHeight <<
+                ")) from " << textureSpec.width << " x " << textureSpec.height << " texture at mip level " << textureSpec.level <<
+                " to " << bufferSpec.size << "-byte buffer with offset " << bufferSpec.offset << " and row pitch " << bufferSpec.rowPitch << std::endl;
+        }
+
+
+};
+
+class CopyTests_B2T : public CopyTests {
+protected:
+    static void FillBufferData(RGBA8* data, size_t count) {
+        for (size_t i = 0; i < count; ++i) {
+            data[i] = RGBA8(
+                static_cast<uint8_t>(i % 255), 
```

Comment:
Did you mean to do all of that with 256 instead? Here the 256th element will have (1, 1, 0, 0) wich looks like 257.